File: 23. Merge k Sorted Lists/23.cpp
class Solution
{
public:
    ListNode *mergeKLists(vector<ListNode *> &lists)
    {
        //基本情况
        if (lists.size() == 0)
            return nullptr;
        //大小堆
        priority_queue<int, vector<int>, greater<int>> q;
        for (auto l : lists)
        {
            while (l)
            {
                q.push(l->val);
                l = l->next;
            }
        }
        ListNode *head = new ListNode(0);
        ListNode *p = head;
        while (!q.empty())
        {
            p->next = new ListNode(q.top());
            p = p->next;
            q.pop();
            if (p->next)
                q.push(p->next->val);
        }
        return head->next;
    }
};

File: 23. Merge k Sorted Lists/23.kt
class Solution {
    fun mergeKLists(lists: Array<ListNode?>): ListNode? {
        if (lists.isEmpty()) return null
        val dummy = ListNode(-1)
        var p: ListNode? = dummy
        val pq = PriorityQueue(
            lists.size
        ) { a: ListNode?, b: ListNode? -> a!!.`val` - b!!.`val` }
        for (head in lists) {
            if (head != null) pq.add(head)
        }
        while (!pq.isEmpty()) {
            val node = pq.poll()
            p!!.next = node
            if (node!!.next != null) {
                pq.add(node.next)
            }
            p = p.next
        }
        return dummy.next
    }
}